### ISSUE TITLE
SOURCE is missing

### ISSUE THREAD

**Issue by jyapujuju** [probonopd/AppImageKit#381]
./appimagetool-x86_64.AppImage
:::system::: 
ubuntu 16.04.2 x64

**Comment by probonopd** [comment]
Hi @jyapujuju what did you provide as the source argument?

**Comment by jyapujuju** [comment]
im on 64 bit machine main objective was to make games run on 64 bit machines. but i provide atom.io as a try.

**Comment by probonopd** [comment]
You need to provide an AppDir as the `<SOURCE>` argument, like so:

```
./appimagetool-x86_64.AppImage ./TheSourceAppDir/
```

**Comment by philippgille** [comment]
Maybe you could print the example usage when calling `appimagetool-x86_64.AppImage` without any arguments.

You've got it in the README of the repository:

```
Usage:
  appimagetool [OPTION...] SOURCE [DESTINATION] - Generate, extract, and inspect AppImages

Help Options:
  -h, --help                  Show help options

Application Options:
  -l, --list                  List files in SOURCE AppImage
  -u, --updateinformation     Embed update information STRING; if zsyncmake is installed, generate zsync file
  --bintray-user              Bintray user name
  --bintray-repo              Bintray repository
  --version                   Show version number
  -v, --verbose               Produce verbose output
  -s, --sign                  Sign with gpg2
  -n, --no-appstream          Do not check AppStream metadata
```

**Comment by TheAssassin** [comment]
I agree, @philippgille. Very good idea. PRs welcome.

**Comment by philippgille** [comment]
I'm not a C programmer, but for anyone interested:

[https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L299](https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L299):

`die("SOURCE is missing");`

And a few lines before that: [https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L272](https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L272)

`if (!g_option_context_parse (context, &argc, &argv, &error))`

Which is where the arguments get parsed. So either check the `&argv` length (I guess?) right before that and print the help text if it's 0 and die (like `if (...) { die("<helptext>"); }`. Or maybe even do that in the first line of the `main` method: [https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L244](https://github.com/AppImage/AppImageKit/blob/appimagetool/master/appimagetool.c#L244)

**Comment by probonopd** [comment]
Well, it should be sufficient to replace

```
die("SOURCE is missing");
````

with code that prints the usage statement. But how do I do this with https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html?

**Comment by TheAssassin** [comment]
Quite easy. Just print the usage details before the die call. It can remain in there to give the user a hint why the application refuses to work. Most other software does that, too, when printing the usage details by default.

I'd suggest to add `void printUsage() {...}`, and call that when either `-h` or no argument is passed.

**Comment by probonopd** [comment]
Yes, but how can I get the automatically generated (by https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html) usage statement to print?

**Comment by TheAssassin** [comment]
Ah. I thought it was sufficient to just print the message @philippgille found in the README, hardcoded in the source code.

>Another important feature of GOption is that it can automatically generate nicely formatted help output. Unless it is explicitly turned off with g_option_context_set_help_enabled(), GOption will recognize the --help, -?, --help-all and --help-groupname options (where groupname is the name of a GOptionGroup) and write a text similar to the one shown in the following example to stdout.
(source: https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html)

Sounds like the easiest way is to just insert a `-h` when no argument is passed.

Hacker's solution:
```
if(argc <= 1)
{
  argc++;
  char *argv_[2];
  argv_[0] = argv[0];
  argv_[1] = "-h";
  argv = argv_;
}

// call argument parser afterwards...
```

**Comment by probonopd** [comment]
This does not work:

```
linux@linux:~>  /home/linux/Downloads/appimagetool-x86_64.AppImage 

(process:7629): GLib-CRITICAL **: g_path_get_basename: assertion 'file_name != NULL' failed
Segmentation fault (core dumped)
```

We need to either also pass in a dummy arg (which I am trying now) or find out how to properly get https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html to print a usage statement...
